### ISSUE TITLE
DrawLine line thickness different from XAML line thickness

### ISSUE THREAD

**Issue by octav84** [Microsoft/Win2D#404]
Hi
I draw a table on a CanvasVirtualControl. This table has a header which is drawn using a XAML ListView. The cells drawn by the XAML listView as well as the the cells drawn using Win2D have a border of thickness = 1. But when actually rendering the cells, the border line drawn with Win2D is twice as thick as the one drawn by the XAML Border control. Actually, drawing is even more inconsistent: the cells drawn with Win2D have Top, Right and Bottom border, all having thickness of 1 pixel. The Right and Bottom borders look the same, and the Top border is twice thicker than them, and 3 times thicker than the border drawn with XAML. So the lines drawn with Win2D are not even all the same, although I set the same thickness for all of them.

![cell drawing issue](https://cloud.githubusercontent.com/assets/3714293/17395463/2c86583a-5a37-11e6-9f27-c0e96d26a8c6.png)

 I attached a picture to illustrate.
The first row in the table is drawn using only XAML controls. The rows under (except the first column on the left, the one with the type names for the values in the table) are drawn using Win2D. I should be obvious that the separator lines thickness varies - although it was set to the same value of 1 pixel.

**Comment by shawnhar** [comment]
> although it was set to the same value of 1 pixel

Do you really mean one pixel (did you explicitly change your unit mode to operate in pixels?) or 1 dip?  Win2D operates in dips by default (see http://microsoft.github.io/Win2D/html/DPI.htm)

Next question is what DPI is your app running at?  The most likely reason for a 1-dip-wide line to cover more than one pixel is if your DPI is > 96.

Another possibility is if you are drawing at non-integer pixel locations, the line may be getting antialiased.  To draw repeated elements such as a grid with consistent sizing for each, you will need to either snap to integer pixel locations, or disable antialiasing.

**Comment by octav84** [comment]
Thanks for the quick response. 
I don't use actual pixels, I meant "1 unit" - let's say 1 DIP.
Second, I do use non-integer locations for the X coordinates. I believe I cannot change the actual coordinates to snap to integer locations - due to some weird compliance issue with a XAML - drawn header. Could you tell me how to disable the antialising ?

**Comment by shawnhar** [comment]
> I believe I cannot change the actual coordinates to snap to integer locations - due to some weird compliance issue with a XAML - drawn header

Sure you can.  You're the one doing this drawing - you can round whatever values you need while doing so.

XAML rendering does similar things internally, snapping rendering positions to integer pixel locations, for basically the same reason.


> Could you tell me how to disable the antialising ? 

http://microsoft.github.io/Win2D/html/P_Microsoft_Graphics_Canvas_CanvasDrawingSession_Antialiasing.htm

**Comment by octav84** [comment]
Ok, after almost an entire day of trial and error, I found that, for line thickness greater than 1 DIP, both XAML and Win2D generate identical lines, both in thickness and in color. 
For a line of exactly 1 DIP, Win2D generates a line twice as thick than XAML does, which is also a whiter shade than it should. For a set code of #D9D9D9, the actual measured color usingInstant Eye Dropper tool is #E7E7E7 - see my related issue here: [https://github.com/Microsoft/Win2D/issues/403](url)
Using the same tool I can also clearly see that the line generated by Win2D to be twice as thick as the one generated by XAML, and it also has a whiter gradient.

I fixed it by disabling Antialising, as you said, and by setting CanvasStrokeStyle.TransformBehaviour to Hairline, for 1 DIP lines.
